### DISCUSSION TITLE
HomeBridge "Blocked" by njsPC

### DISCUSSION THREAD

**Discussion by arrmo** [tagyoureit/nodejs-poolController#597] [App and O/S Config Questions] [answered]
Hi,

Just to let you know, and issue I think I'm seeing ... 😆. Not sure if this is the right place to flag it or not, but it seemed to show up when I moved to v7.7.0.

I have connected njsPC to HomeBridge, using https://github.com/gadget-monk/homebridge-poolcontroller. It has worked well for a long time, but just recently I noticed issues => coincident with my connectivity issues, that I can see, but also ... when njsPC is still initializing, it seems to fully take HomeBridge down (for all devices). Is this expected? I'm thinking not ... LOL.

Thanks!

**Comment by rstrouse** [comment]
Do you by chance have the logging level set high on homebridge.  I remember seeing some older posts regarding the logging level slowing down the plugin.   Also, if this is happening during the time EasyTouch is reporting its data, it can potentially send more socket data to homebridge but I wouldn't think that is burdensome to homebridge.

**Reply by tagyoureit** [reply]
That logging issue was for Hubitat but I could see it affecting HomeBridge in the same way. 

@arrmo do you have this setting enabled?

> IgnoreControllerReadyState: Normally the plugin waits for poolController to report "Ready" status, setting this to true disables this check. Not Recommended for normal use.

This would cause excessive load under startup conditions.

**Reply by arrmo** [reply]
> Do you by chance have the logging level set high on homebridge. I remember seeing some older posts regarding the logging level slowing down the plugin. Also, if this is happening during the time EasyTouch is reporting its data, it can potentially send more socket data to homebridge but I wouldn't think that is burdensome to homebridge.

Sorry for the slow reply! Got tied up with other home related repairs 🤣. No, no special HomeBridge settings. Assuming this would not be the default, agreed?



> > IgnoreControllerReadyState: Normally the plugin waits for poolController to report "Ready" status, setting this to true disables this check. Not Recommended for normal use.

Sorry, to understand - are you recommending to set this value (true)? Thanks!

**Reply by tagyoureit** [reply]
IgnoreControllerReadyState should be set to false.  You want to respect the loading state and not have the homebridge controller flooding with startup config messages.

**Reply by arrmo** [reply]
Changed that setting (not sure why it was set, seemed to be the default) => all seems OK now. Thanks!

**Reply by arrmo** [reply]
@tagyoureit, not sure how to mark your comment as answered - but will if I can 😆

**Reply by arrmo** [reply]
OK, failing again! I got njsPC all back up, and now HomeBridge is down ... LOL. Here is what I see in the HB log,
```
[9/19/2022, 4:39:06 PM] Homebridge v1.5.0 (HAP v0.10.2) (Homebridge E946) is running on port 51970.
[9/19/2022, 4:39:06 PM] [PoolControllerPlatform] Version of poolController 1.0 does not meet the minimum for this Homebridge plugin 6.0
[9/19/2022, 4:39:06 PM] [HB Supervisor] Homebridge Process Ended. Code: 0, Signal: null
```

Thoughts?

Thanks!

**Reply by rstrouse** [reply]
Which repo are you using is it the one from @gadget-monk?  I think he misunderstood what the version is that is returned in the xml for upnp.  That is the spec version for upnp for which we are using 1.0.  To be fair I think at one point we incorrectly put the major, minor, and patch for njsPC in there.  This would cause clients that didn't support the spec to ignore us.

I added an appVersion node to the xml to get the version.  This will require some changes to the index.js on the homebridge plugin.

Lines 127 & 128 should read as follows.
```
        var major = parseInt(body.match(/<device>[\s|\S]+<appVersion>[\s|\S]+<major>(\d+)<\/major>/)[1], 10);
        var minor = parseInt(body.match(/<device>[\s|\S]+<appVersion>[\s|\S]+<minor>(\d+)<\/minor>/)[1], 10);
```

**Reply by arrmo** [reply]
> Which repo are you using is it the one from @gadget-monk?

Yes! Back when I set this up (a while ago!) it was the only one ... is it the "wrong" one?



> Lines 127 & 128 should read as follows.

Made those changes, a bit different now,
```
[9/19/2022, 7:25:07 PM] [PoolControllerPlatform] Initializing PoolControllerPlatform platform...
[9/19/2022, 7:25:07 PM] [PoolControllerPlatform] Loading PoolControllerPlatform
...
[9/19/2022, 7:25:07 PM] [PoolControllerPlatform] Starting UPnP search for PoolController.
```
And ... HomeBridge restarts again. I know it's due to this plugin => disable it, no issues at all.

Thanks!

**Reply by rstrouse** [reply]
So is it working or is it stuck searching ssdp packets for PoolController?

**Reply by rstrouse** [reply]
I don't know how to get the logs from this plugin but if you increase the logging level to 4 it should tell us what is giving it IBS.

**Reply by gadget-monk** [reply]
i don't use homebridge anymore so i dont have a way to test, but if the problem is upnp, try entering the hostname/IP and port of njsPC and see if that fixes the problem. I can edit the repo to change the two lines.

**Reply by arrmo** [reply]
That does make a difference! When I use `*` for the server and port, that's where I see the njsPC logs filling with,
```
1|njsPC    | [9/18/2022, 8:33:50 PM] info: [8:33:50 PM] 192.168.2.64 GET /upnp.xml/device {}
1|njsPC    | [9/18/2022, 8:33:56 PM] info: [8:33:56 PM] 192.168.2.64 GET /upnp.xml/device {}
1|njsPC    | [9/18/2022, 8:34:02 PM] info: [8:34:02 PM] 192.168.2.64 GET /upnp.xml/device {}
1|njsPC    | [9/18/2022, 8:34:09 PM] info: [8:34:09 PM] 192.168.2.64 GET /upnp.xml/device {}
```

So this is the mystery from the other day 😆. And ... each line above is when Homebridge restarts. Disable this plug in, and the HB restarts end (and no longer the info above in the njsPC log).

So then, setting the server to `http://myserver:4200` ... HB still keeps restarting, this time the njsPC log shows,
```
1|njsPC  | [9/19/2022, 8:59:37 PM] info: [8:59:37 PM] 192.168.2.64 GET /device {}
1|njsPC  | [9/19/2022, 8:59:49 PM] info: [8:59:49 PM] 192.168.2.64 GET /device {}
1|njsPC  | [9/19/2022, 9:00:00 PM] info: [9:00:00 PM] 192.168.2.64 GET /device {}
```
Again, each row is when HB restarts ... never getting fully up though. And, in the HB log,
```
[9/19/2022, 9:00:44 PM] [PoolControllerPlatform] Error fetching data: Request failed with status code 404
[9/19/2022, 9:00:44 PM] Error: Request failed with status code 404
    at createError (/usr/lib/node_modules/axios/lib/core/createError.js:16:15)
    at settle (/usr/lib/node_modules/axios/lib/core/settle.js:17:12)
    at IncomingMessage.handleStreamEnd (/usr/lib/node_modules/axios/lib/adapters/http.js:293:11)
    at IncomingMessage.emit (node:events:402:35)
    at endReadableNT (node:internal/streams/readable:1343:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21)
```
I did check, and going to `http://myserver:4200/device` does in fact return,
```
Cannot GET /device
```
Thinking the url has changed? For example, I can get to `http://myserver:4200/state/all`, or even `http://myserver:4200/state/device`, though it returns,
```
{}
```
Not sure quite what you are looking to return?

Thanks!

**Reply by rstrouse** [reply]
Change line 689 to
```
opts.url = `${poolURL}/upnp.xml`
```

Change line 95 to
```
self.config.ip_address = headers.LOCATION.replace('/upnp.xml', '');
```

As for the http://myserver/state/all it should return the current state for all equipment on the pool.  That is if the myserver:4200 is pointing to a valid njsPC instance.  And `http://myserver:4200/upnp.xml` should return the upnp device definition.

**Reply by arrmo** [reply]
> As for the http://myserver/state/all it should return the current state for all equipment on the pool. That is if the myserver:4200 is pointing to a valid njsPC instance. And `http://myserver:4200/upnp.xml` should return the upnp device definition.

Yes, both do - thanks!

Made the changes you noted (and kept the earlier changes above ... right?). Still causes HB grief - odd. As soon as I enable the plugin, HB reset cycles, and shows,
```
[9/20/2022, 8:05:16 PM] Homebridge v1.5.0 (HAP v0.10.2) (Homebridge E946) is running on port 51970.
[9/20/2022, 8:05:16 PM] TypeError: Cannot read properties of null (reading '1')
    at PoolControllerPlatform.validateVersion (/usr/lib/node_modules/homebridge-poolcontroller/index.js:130:97)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
[9/20/2022, 8:05:16 pm] [SmartThings-v2] ERROR: uncaughtException: Cannot read properties of null (reading '1')
TypeError: Cannot read properties of null (reading '1')
    at PoolControllerPlatform.validateVersion (/usr/lib/node_modules/homebridge-poolcontroller/index.js:130:97)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) 
[9/20/2022, 8:05:16 PM] Got SIGTERM, shutting down Homebridge...
```

Let me check those noted lines, try some debugging. Thanks again.

**Reply by arrmo** [reply]
OK, it's the code related to version. I hacked in for now,
```
    var valid = false
        , validMajor = 1
        , validMinor = 0
```

Just to let the code proceed => it does, HB stops restarting, and I tried the lights in HomeKit - they work! So your last 2 changes above, and also the version seem to get it up and running. Now to figure out that version code 😆.

Thanks!

**Reply by arrmo** [reply]
OK, I think I see the issue - the code you sent for appVersion ... that isn't a field (or whatever the right word is 😆) in the xml output. Agreed?

**Reply by rstrouse** [reply]
Element is the right word and if you pull njsPC it will now include it in the XML.
```
<root xmlns="urn:schemas-upnp-org:device-1-0">
<specVersion>
   <major>1</major>
   <minor>0</minor>
</specVersion>
<device>
   <deviceType>urn:schemas-tagyoureit-org:device:PoolController:1</deviceType>
   <friendlyName>NodeJS Pool Controller</friendlyName>
   <manufacturer>tagyoureit</manufacturer>
   <manufacturerURL>https://github.com/tagyoureit/nodejs-poolController</manufacturerURL>
   <presentationURL>http://192.168.1.234:4200/state/all</presentationURL>
   <appVersion>
      <major>7</major>
      <minor>7</minor>
      <patch>0</patch>
   </appVersion>
   <modelName>njsPC v7.7.0</modelName>
   <modelNumber>njsPC7-7-0</modelNumber>
   <modelDescription>An application to control pool equipment.</modelDescription>
   <serialNumber>10:7b:44:f1:8e:21</serialNumber>
   <UDN>uuid:806f52f4-1f35-4e33-9299-107b44f18e21::urn:schemas-tagyoureit-org:device:PoolController:1</UDN>
   <serviceList/>
   <deviceList/>
</device>
</root>
```

**Reply by arrmo** [reply]
Yep, that works! Pulled it - checked in a browser (XML), and it's got the appVersion now, as you say. And HB - can enable it now, no restarts 😄. Thanks!

How to get the corrections in to the plugin?

**Reply by gadget-monk** [reply]
I can update the plug-in code. Which lines need to be changed?

**Reply by arrmo** [reply]
Pretty sure I know the 4 lines, but ... is there a clean way to check the diff? As I modified it locally, not sure how to confirm that.

Thanks!

**Comment by rstrouse** [comment, marked as answer]
Lines 128 & 129 should read
```
 var major = parseInt(body.match(/<device>[\s|\S]+<appVersion>[\s|\S]+<major>(\d+)<\/major>/)[1], 10);
 var minor = parseInt(body.match(/<device>[\s|\S]+<appVersion>[\s|\S]+<minor>(\d+)<\/minor>/)[1], 10);
```
Line 95
`self.config.ip_address = headers.LOCATION.replace('/upnp.xml', ''); `

Line 689
```
opts.url = `${poolURL}/upnp.xml` 
  ```

**Reply by gadget-monk** [reply]
thanks! i pushed the updates out on github

**Reply by arrmo** [reply]
Seems to work great, thanks!

**Reply by kurtifier** [reply]
Worked for me as well! Thank you @gadget-monk and @arrmo !!

**Reply by racerx9146** [reply]
works for me as well. You guys rock! If you are ever in Rocklin, California beers are on me :) @gadget-monk @gadget-monk @kurtifier

**Reply by rstrouse** [reply]
Funny I'm in Roseville.

**Reply by gadget-monk** [reply]
All thanks to @rstrouse and @arrmo (small world though, I am in the Bay Area)

**Comment by cjkeithley** [comment]
And I’m in Loomis!

**Reply by rstrouse** [reply]
All on the I80 corridor

**Reply by arrmo** [reply]
Dang it, I'm over in Texas ... LOL. Actually, transplanted Canuck 🍁 

Glad this is working for folks, it really is a great feature!

**Reply by racerx9146** [reply]
it is a small world. 
I know not everyone is a Apple fan but its really nice to use HomeKit turn on spa when away from home or check if things are on/off. Did not really want to open my firewall or vpn to get remote access to the pool controller dashboard. And it is very fun to say "Siri, turn on the waterfall"

**Reply by rstrouse** [reply]
"Siri call me an ambulance"
"Ok from now on I will call you ... an Amulance"

**Reply by arrmo** [reply]
> it is a small world. I know not everyone is a Apple fan but its really nice to use HomeKit turn on spa when away from home or check if things are on/off. Did not really want to open my firewall or vpn to get remote access to the pool controller dashboard. And it is very fun to say "Siri, turn on the waterfall"

And set the temperature ... when my butt is in the spa, and the spa side remote is fried (due to weather) 😆

**Reply by tagyoureit** [reply]
I'm in Bay Area, too.  Los Gatos.

**Reply by DrJeff** [reply]
Don't forget me I didn't help but I'm in San Jose..... lol
